### ISSUE TITLE
`flutter test` doesn't break on exceptions when running under debugger (since dart 2)

### ISSUE THREAD

**Issue by DanTup** [flutter/flutter#16741]
If I create a test that throws an exception and run with `flutter run --machine --start-paused` I would expect it to stop on exceptions, but it doesn't seem to. I think it did in the past (maybe the first flutter beta?) because I had Travis test failures because of it pausing; but today it seems to just continue and spit the error to the console.

**Comment by DanTup** [comment]
Did some testing of older versions:

- Flutter v0.1.5 - ✅ breaks
- Flutter v0.1.8 - ✅ breaks
- Flutter v0.2.3 - ✅ breaks
- Flutter v0.2.8 - ❌ does not break
- Flutter v0.3.1 - ❌ does not break

So, this seems like a regression. I'll see if I can narrow it down.

**Comment by zoechi** [comment]
Have you tried 0.2.4 and later with `--no-preview-dart-2`?

**Comment by DanTup** [comment]
Git bisect points at 71b2cfb2a0724af692c298a62f86b3a934b6d3ca which is enabling Dart 2 (which I guess is @zoechi's hunch)..

**Comment by DanTup** [comment]
Indeed, `--no-preview-dart-2` on v0.2.4 restores the behaviour. (cc @mit-mit @devoncarew - I guess this issue is outside of flutter tools, not sure if there's someone appropriate to ping?).

**Comment by DanTup** [comment]
Actually, this repros in standard Dart with `package:test` if you pass `--preview-dart-2` too, so maybe unrelated to flutter. Not sure whether it's Dart VM or maybe `package:test`. I can re-raise elsewhere if someone can advise which it's likely to be.

**Comment by zoechi** [comment]
Breakpoints work for me in IntelliJ with `flutter run` and `flutter test`
```
Flutter 0.3.5 • channel dev • https://github.com/flutter/flutter.git
Framework • revision 7ffcd3d22d (13 hours ago) • 2018-04-24 14:03:41 -0700
Engine • revision ec611470b5
Tools • Dart 2.0.0-dev.48.0.flutter-fe606f890b
```

**Comment by DanTup** [comment]
@zoechi It's unhandled exceptions that aren't working. If you put `throw new Exception('test');` inside a test and run under the debugger, it just spits the exception text out in the output but doesn't pause the debugger (for me at least). This means you then need to go back and add breakpoints manually.

A minor annoyance, sure, but I'm writing automated tests for Dart Code's debug adapter atm so I have to chose between skipping the tests or writing them wrong for now :D

**Comment by zoechi** [comment]
Sorry, didn't pay attention.

**Comment by devoncarew** [comment]
@sivachandra @sivachandra @aam, it looks like the behavior of breaking on exceptions has regressed with `--preview-dart-2`.

**Comment by rmacnak-google** [comment]
Looks like  kernel::FlowGraphBuilder::CatchBlockEntry is labeling this catch clause as not generated, so the VM considers the exception handled.

**Comment by alexmarkov** [comment]
Related dart-sdk issue: https://github.com/dart-lang/sdk/issues/29056

**Comment by aam** [comment]
The dart sdk fix should have landed in Flutter b2c98f9a4eb5eef85e0c69b99713beb4ceba02ca

**Comment by DanTup** [comment]
Fix looks good, thanks! 👍
